hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java

```diff
@@ -274,11 +275,14 @@ private boolean handleEofException(Exception e, WALEntryBatch batch) {
     // since we don't add current log to recovered source queue so it is safe to remove.
     if ((e instanceof EOFException || e.getCause() instanceof EOFException) &&
       (source.isRecovered() || queue.size() > 1) && this.eofAutoRecovery) {
-      Path head = queue.peek();
+      Path path = queue.peek();
       try {
-        if (fs.getFileStatus(head).getLen() == 0) {
-          // head of the queue is an empty log file
-          LOG.warn("Forcing removal of 0 length log in queue: {}", head);
+        if (!fs.exists(path)) {
+          // There is a chance that wal has moved to oldWALs directory, so look there also.
+          path = AbstractFSWALProvider.getArchivedLogPath(path, conf);
```

Comment:
Oh, this is not very stable...

You need to try all the possible path, just like what we have done in WALEntryStream. So I suggest we move the main logic of WALEntryStream.getArchivedLog to AbstractFSWALProvider, as a util method, and we call this method in both places.

Reply:
Thank you @Apache9  for the feedback. I added a new util method in AbstractFSWALProvider to look into both the oldWALs path if it can't find the log in WALs directory.
After the change, there are 2 methods in `AbstractFSWALProvider` now to get archiveLog path:
1.  `getArchivedLog` This is the method added by this patch.
2.  `getArchivedLogPath` This was already present which looks into separate oldLogDir path based on config.

I think we can remove `getArchivedLogPath` method and replace all callers with getArchivedLog. WDYT ?

Reply:
Oh, just saw this comment. Please see my new comment below, you need to keep getArchivedLogPath, as it is used when we want to archive a wal file.
It is OK to remove the getArchivedLog in WALEntryStream.